grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -435,17 +438,78 @@ private static String getResponseBody(List<FieldDescriptor> topLevelFields,
         }
-        for (FieldDescriptor fieldDescriptor: topLevelFields) {
+        for (FieldDescriptor fieldDescriptor : topLevelFields) {
             if (fieldDescriptor.getName().equals(responseBody)) {
                 return responseBody;
             }
         }
         return null;
     }
 
+    private static MediaType getMediaTypeFromHttpBody(JsonNode jsonNode) {
+        final String contentType = jsonNode.get("contentType").asText();
+        try {
+            return MediaType.parse(contentType);
+        } catch (IllegalArgumentException e) {
+            logger.warn("Invalid media type in http body content_type {}.", contentType);
+            return MediaType.JSON_UTF_8;
+        }
+    }
+
+    private static Function<HttpData, MediaType> generateMediaTypeDecider(TranscodingSpec spec) {
+        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getOutputType())) {
+            return httpData -> {
+                try (HttpData data = httpData) {
+                    final byte[] array = data.array();
+                    try {
+                        final JsonNode jsonNode = mapper.readValue(array, JsonNode.class);
+                        return getMediaTypeFromHttpBody(jsonNode);
+                    } catch (IOException e) {
+                        logger.warn("Unexpected exception while reading HttpBody {}.", data, e);
+                    }
+
+                    // Default behavior is to return JSON_UTF_8, the default content type for JSON transcoding.
+                    return MediaType.JSON_UTF_8;
+                }
+            };
+        }
+
+        return httpData -> MediaType.JSON_UTF_8;
+    }
+
     @Nullable
     private static Function<HttpData, HttpData> generateResponseBodyConverter(TranscodingSpec spec) {
-        @Nullable final String responseBody = spec.responseBody;
+        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getOutputType())) {
+            return httpData -> {
+                try (HttpData data = httpData) {
+                    final byte[] array = data.array();
+
+                    final JsonNode jsonNode;
+                    try {
+                        jsonNode = mapper.readValue(array, JsonNode.class);
+                    } catch (IOException e) {
+                        logger.warn("Unexpected exception while parsing HttpBody from {}", data, e);
+                        return HttpData.wrap(array);
+                    }
+
+                    final String httpBody = jsonNode.get("data").asText();
+                    final MediaType mediaType = getMediaTypeFromHttpBody(jsonNode);
+                    final Charset charset = mediaType.charset();
+
+                    final byte[] httpBodyBytes;
+                    if (charset != null) {
+                        httpBodyBytes = Base64.getDecoder().decode(httpBody);
+                    } else {
+                        httpBodyBytes = Base64.getDecoder().decode(httpBody);
+                    }
```

Comment:
Response is Base64 because protobuf representation of body is `byte[]` and jackson serializes it as a base64 string.

Reply:
I assume performance is not our first concern here so it seems fine 🙂